# Corn Data Cleaning Notebook

Data for Grain Corn for the years 1950 to 2020 was downloaded from https://quickstats.nass.usda.gov/. This website restricts data downloads to 5,000 rows of data at a time. Thus, I was required to download data for one year at a time. This resulted in over 70 csv files being downloaded. These csv files are stored in the `Corn Data` folder. I will import them them all in this notebook, clean the data, combine it all into a single data frame and export it as a csv for further analysis.

In [1]:
# Import tools 
import pandas as pd
import numpy as np
import os

I will now import all of the data as individual dataframes. I will store these data frames in a list and combine them later.

In [2]:
# I will need a list for storing the individual data frames made from the csv files
corn_list = []

# read through all files in the Corn Data folder and add to corn_list
for filename in os.listdir("./Corn Data"):
    
    # import data from each file as a data frame
    file_data = pd.read_csv(f"./Corn Data/{filename}")
    
    # add the data frames to the list
    corn_list.append(file_data)  

# confirm that the data frames were created successfully. 
# Since this is a simple procedure, I will only check that one was created successfully.
display(corn_list[50].head())

C:\Users\Daniel\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
C:\Users\Daniel\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (19) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
C:\Users\Daniel\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (7,9,12,19,20) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
C:\Users\Daniel\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (19,20) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cel

Program  Year          Period Week Ending Geo Level       State  State ANSI  \
0  SURVEY  2000  MARKETING YEAR         NaN  NATIONAL    US TOTAL         NaN   
1  SURVEY  2000  MARKETING YEAR         NaN     STATE     ALABAMA         1.0   
2  SURVEY  2000  MARKETING YEAR         NaN     STATE     ARIZONA         4.0   
3  SURVEY  2000  MARKETING YEAR         NaN     STATE    ARKANSAS         5.0   
4  SURVEY  2000  MARKETING YEAR         NaN     STATE  CALIFORNIA         6.0   

  Ag District  Ag District Code County  ...  Zip Code  Region  watershed_code  \
0         NaN               NaN    NaN  ...       NaN     NaN               0   
1         NaN               NaN    NaN  ...       NaN     NaN               0   
2         NaN               NaN    NaN  ...       NaN     NaN               0   
3         NaN               NaN    NaN  ...       NaN     NaN               0   
4         NaN               NaN    NaN  ...       NaN     NaN               0   

   Watershed  Commodity                                         Data Item  \
0        NaN       CORN  CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
1        NaN       CORN  CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
2        NaN       CORN  CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
3        NaN       CORN  CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
4        NaN       CORN  CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   

  Domain Domain Category Value CV (%)  
0  TOTAL   NOT SPECIFIED  1.85    NaN  
1  TOTAL   NOT SPECIFIED  2.16    NaN  
2  TOTAL   NOT SPECIFIED  2.78    NaN  
3  TOTAL   NOT SPECIFIED  1.75    NaN  
4  TOTAL   NOT SPECIFIED  2.44    NaN  

[5 rows x 21 columns]

The csv files were read in successfully. I will now merge them, or concatenate them, into a single data frame.

In [3]:
# concat all the dataframes into 1.
# make sure to reassign index values so there aren't duplicate values
corn_df = pd.concat(corn_list, ignore_index=True)

# confirm that the data frame was created successfully
display(corn_df.head())

Program  Year          Period Week Ending Geo Level     State  State ANSI  \
0  SURVEY  1950  MARKETING YEAR         NaN  NATIONAL  US TOTAL         NaN   
1  SURVEY  1950  MARKETING YEAR         NaN     STATE   ALABAMA         1.0   
2  SURVEY  1950  MARKETING YEAR         NaN     STATE   ARIZONA         4.0   
3  SURVEY  1950  MARKETING YEAR         NaN     STATE   ARIZONA         4.0   
4  SURVEY  1950  MARKETING YEAR         NaN     STATE  ARKANSAS         5.0   

  Ag District  Ag District Code County  ...  Zip Code  Region watershed_code  \
0         NaN               NaN    NaN  ...       NaN     NaN              0   
1         NaN               NaN    NaN  ...       NaN     NaN              0   
2         NaN               NaN    NaN  ...       NaN     NaN              0   
3         NaN               NaN    NaN  ...       NaN     NaN              0   
4         NaN               NaN    NaN  ...       NaN     NaN              0   

   Watershed Commodity                                          Data Item  \
0        NaN      CORN   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
1        NaN      CORN   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
2        NaN      CORN   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
3        NaN      CORN  CORN, SILAGE - PRICE RECEIVED, MEASURED IN $ /...   
4        NaN      CORN   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   

  Domain Domain Category Value CV (%)  
0  TOTAL   NOT SPECIFIED  1.52    NaN  
1  TOTAL   NOT SPECIFIED  1.48    NaN  
2  TOTAL   NOT SPECIFIED  2.04    NaN  
3  TOTAL   NOT SPECIFIED   7.2    NaN  
4  TOTAL   NOT SPECIFIED  1.42    NaN  

[5 rows x 21 columns]

The data frame was created successfully.

## Corn Dataframe EDA and cleaning

I will now do some exploratory data analysis to get to know the corn data.

In [4]:
corn_df.shape

(1873192, 21)

1.87 million rows and 21 columns. I will not need all of this data.

In [5]:
corn_df.head(5)

Program  Year          Period Week Ending Geo Level     State  State ANSI  \
0  SURVEY  1950  MARKETING YEAR         NaN  NATIONAL  US TOTAL         NaN   
1  SURVEY  1950  MARKETING YEAR         NaN     STATE   ALABAMA         1.0   
2  SURVEY  1950  MARKETING YEAR         NaN     STATE   ARIZONA         4.0   
3  SURVEY  1950  MARKETING YEAR         NaN     STATE   ARIZONA         4.0   
4  SURVEY  1950  MARKETING YEAR         NaN     STATE  ARKANSAS         5.0   

  Ag District  Ag District Code County  ...  Zip Code  Region watershed_code  \
0         NaN               NaN    NaN  ...       NaN     NaN              0   
1         NaN               NaN    NaN  ...       NaN     NaN              0   
2         NaN               NaN    NaN  ...       NaN     NaN              0   
3         NaN               NaN    NaN  ...       NaN     NaN              0   
4         NaN               NaN    NaN  ...       NaN     NaN              0   

   Watershed Commodity                                          Data Item  \
0        NaN      CORN   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
1        NaN      CORN   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
2        NaN      CORN   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   
3        NaN      CORN  CORN, SILAGE - PRICE RECEIVED, MEASURED IN $ /...   
4        NaN      CORN   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU   

  Domain Domain Category Value CV (%)  
0  TOTAL   NOT SPECIFIED  1.52    NaN  
1  TOTAL   NOT SPECIFIED  1.48    NaN  
2  TOTAL   NOT SPECIFIED  2.04    NaN  
3  TOTAL   NOT SPECIFIED   7.2    NaN  
4  TOTAL   NOT SPECIFIED  1.42    NaN  

[5 rows x 21 columns]

There are 21 columns in this data frame. The columns of interest will be:
- `Year`
- `Period`
- `Geo Level`
- `Data Item`
- `Value`

For simplicity, I will drop all of the other columns.

In [6]:
# set conr_df equal to the corn_df with just the columns of interest
corn_df = corn_df[['Year', 'Period', 'Geo Level', 'Data Item', 'Value']]

#confirm drop successfully
display(corn_df.head())

Year          Period Geo Level  \
0  1950  MARKETING YEAR  NATIONAL   
1  1950  MARKETING YEAR     STATE   
2  1950  MARKETING YEAR     STATE   
3  1950  MARKETING YEAR     STATE   
4  1950  MARKETING YEAR     STATE   

                                           Data Item Value  
0   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU  1.52  
1   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU  1.48  
2   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU  2.04  
3  CORN, SILAGE - PRICE RECEIVED, MEASURED IN $ /...   7.2  
4   CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU  1.42

The data frame has now been simplified significantly. The `Data Item` column contains the labels and the `Value` column contains the values for the data of interest. I will not need all 1.8 million of these data. I will now check what unique values there are to select the values of interest.

In [7]:
display(corn_df['Data Item'].value_counts())

CORN, GRAIN - ACRES HARVESTED                                           192900
CORN, GRAIN - PRODUCTION, MEASURED IN BU                                192450
CORN, GRAIN - OPERATIONS WITH AREA HARVESTED                            181897
CORN, GRAIN - YIELD, MEASURED IN BU / ACRE                              178395
CORN - ACRES PLANTED                                                    154644
                                                                         ...  
CORN, SILAGE, ORGANIC - SALES IN CONVENTIONAL MARKETS, MEASURED IN $         6
CORN, GRAIN, NON-IRRIGATED - EAR COUNT, MEASURED IN STALKS / ACRE            2
CORN, GRAIN, IRRIGATED - EAR COUNT, MEASURED IN STALKS / ACRE                2
CORN - ACRES REMAINING TO BE PLANTED                                         2
CORN, GRAIN - RELATIVE WEIGHT, MEASURED IN PCT OF COMMODITY TOTALS           1
Name: Data Item, Length: 225, dtype: int64

In [8]:
# Count unique data item entries
len(corn_df['Data Item'].unique())

225

There are 225 unique values in this list. I am only interested in seven of these data items:
- `CORN, GRAIN - ACRES HARVESTED`
- `CORN, GRAIN - PRODUCTION, MEASURED IN BU`
- `CORN, GRAIN - YIELD, MEASURED IN BU / ACRE`
- `CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU`
- `CORN, GRAIN - YIELD, MEASURED IN BU / NET PLANTED ACRE`
- `CORN, GRAIN - ACRES PLANTED, NET`
- `CORN, GRAIN - PRODUCTION, MEASURED IN $`

I will throw away all of the other data.

In [9]:
# I need a list of the items I want to keep
data_keep_list = ['CORN, GRAIN - ACRES HARVESTED', 'CORN, GRAIN - PRODUCTION, MEASURED IN BU', 'CORN, GRAIN - YIELD, MEASURED IN BU / ACRE', 'CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU', 'CORN, GRAIN - YIELD, MEASURED IN BU / NET PLANTED ACRE', 'CORN, GRAIN - ACRES PLANTED, NET', 'CORN, GRAIN - PRODUCTION, MEASURED IN $']

# I want to keep the original data frame intact, so I will create a new one to edit. 
corn_df_working = pd.DataFrame()

# iterate through each value in my keep list
for row in data_keep_list:
    
    # select all of the rows from the original data frame that contain the desired "keeper" data
    data_filter = corn_df[corn_df['Data Item'] == row]
    
    # add the "keeper" data to the new data frame
    corn_df_working = pd.concat([corn_df_working, data_filter])

# Confirm that the new data frame was created containing only the data items of interest
print(corn_df_working['Data Item'].value_counts())

# Also, confirm that the data frame looks how I want it to look.
corn_df_working.head(5)

CORN, GRAIN - ACRES HARVESTED                             192900
CORN, GRAIN - PRODUCTION, MEASURED IN BU                  192450
CORN, GRAIN - YIELD, MEASURED IN BU / ACRE                178395
CORN, GRAIN - PRICE RECEIVED, MEASURED IN $ / BU           29189
CORN, GRAIN - YIELD, MEASURED IN BU / NET PLANTED ACRE     23788
CORN, GRAIN - ACRES PLANTED, NET                           22123
CORN, GRAIN - PRODUCTION, MEASURED IN $                     3054
Name: Data Item, dtype: int64


Year Period              Geo Level                      Data Item    Value
50  1950   YEAR  AGRICULTURAL DISTRICT  CORN, GRAIN - ACRES HARVESTED  189,000
53  1950   YEAR  AGRICULTURAL DISTRICT  CORN, GRAIN - ACRES HARVESTED  222,000
56  1950   YEAR  AGRICULTURAL DISTRICT  CORN, GRAIN - ACRES HARVESTED  282,000
59  1950   YEAR  AGRICULTURAL DISTRICT  CORN, GRAIN - ACRES HARVESTED  116,000
62  1950   YEAR  AGRICULTURAL DISTRICT  CORN, GRAIN - ACRES HARVESTED  302,000

The new data frame has all of the data items of interest and appears to have been created properly.

### Modify reporting periods

I now want to make sure that only the reporting periods I am interested in are in the data frame. I want to work with data corresponding to the `Year` and `Month`. Not other periodic data. So, I will drop these data from my table.

In [10]:
print(corn_df_working['Period'].unique())

['YEAR' 'YEAR - AUG FORECAST' 'YEAR - JUN ACREAGE' 'YEAR - NOV FORECAST'
 'YEAR - OCT FORECAST' 'YEAR - SEP FORECAST' 'MARKETING YEAR' 'JAN' 'FEB'
 'MAR' 'APR' 'MAY' 'JUN' 'JUL' 'AUG' 'SEP' 'OCT' 'NOV' 'DEC']


There are for non-year, non-month time periods in the data frame that need to be removed:
- `YEAR - AUG FORECAST`
- `YEAR - JUN ACREAGE`
- `YEAR - NOV FORECAST`
- `YEAR - OCT FORECAST`
- `YEAR - SEP FORECAST`

I'm not clear on whether the `Marketing Year` is refereing to `sweet corn` or `corn for grains`. To simplify the analysis, I'm going to assume it is primarily refering to `sweet corn`. The `Marketing Year` for `sweet corn` begins on January 1, so `Marketing Year` can be converted to year. ([Ref](https://en.wikipedia.org/wiki/Marketing_year))

In [11]:
# Period values to be dropped
time_drop_rows = ['YEAR - AUG FORECAST', 'YEAR - JUN ACREAGE', 'YEAR - NOV FORECAST', 'YEAR - OCT FORECAST', 'YEAR - SEP FORECAST']

# iterate through the items to be dropped.
for row in time_drop_rows:
    
    # select all of the rows from the data frame where the period is equal to the value of interest
    time_filter = corn_df_working[corn_df_working['Period'] == row]
    
    # remove the filtered data from the data frame
    corn_df_working.drop(time_filter.index, inplace=True)

# check taht only year and month values are left in the period column
print(corn_df_working['Period'].unique())

['YEAR' 'MARKETING YEAR' 'JAN' 'FEB' 'MAR' 'APR' 'MAY' 'JUN' 'JUL' 'AUG'
 'SEP' 'OCT' 'NOV' 'DEC']


All of the non-year, non-month periods have been removed. I can now convert marketing year to year.

In [12]:
# select all rows where the period equals marketing year
marketing_filter = corn_df_working['Period'] == 'MARKETING YEAR'

# convert the period in each of these rows to "YEAR"
corn_df_working.loc[marketing_filter, 'Period'] = 'YEAR'

# confirm that the transformation worked successfully
display(corn_df_working["Period"].value_counts())

YEAR    606708
FEB       2190
MAR       2189
MAY       2185
JAN       2184
DEC       2179
APR       2178
SEP       2174
JUL       2172
AUG       2169
JUN       2167
OCT       2143
NOV       2139
Name: Period, dtype: int64

The period column now appears to be filtered properly.

### Remove Unwanted Regions

There are several regions listed in the data.

In [13]:
# show all geographical levels
display(corn_df_working["Geo Level"].value_counts())

COUNTY                          517835
AGRICULTURAL DISTRICT            59100
STATE                            54192
NATIONAL                          1567
AMERICAN INDIAN RESERVATION         72
PUERTO RICO & OUTLYING AREAS         9
REGION : MULTI-STATE                 2
Name: Geo Level, dtype: int64

I don't want to work with every region, just `NATIONAL` data. So, I will slice out only this data.

In [14]:
# select all rows where the geo level is equal to national
marketing_filter = corn_df_working['Geo Level'] == 'NATIONAL'

# set the corn_df_working data frame equal to itself only where the above filter is true
corn_df_working = corn_df_working.loc[marketing_filter]

# check that only the national geo level is retained
display(corn_df_working["Geo Level"].value_counts())

NATIONAL    1567
Name: Geo Level, dtype: int64

In [15]:
# Region values to be dropped
region_drop_rows = ['COUNTY', 'AGRICULTURAL DISTRICT', 'STATE', 'AMERICAN INDIAN RESERVATION', 'PUERTO RICO & OUTLYING AREAS', 'REGION : MULTI-STATE']

# drop the various rows from the
for row in region_drop_rows:
    region_filter = corn_df_working[corn_df_working['Geo Level'] == row]
    corn_df_working.drop(region_filter.index, inplace=True)

print(corn_df_working['Geo Level'].value_counts())

NATIONAL    1567
Name: Geo Level, dtype: int64


The `NATIONAL` geo level is the only one retained at this point. I can thus now drop the `Geo Level` column.

In [16]:
# drop the geo level column
corn_df_working.drop(columns='Geo Level', inplace=True)

# ensure the column was dropped successfully
corn_df_working.head()

Year Period                      Data Item       Value
6422   1950   YEAR  CORN, GRAIN - ACRES HARVESTED  72,398,000
14719  1951   YEAR  CORN, GRAIN - ACRES HARVESTED  71,191,000
22985  1952   YEAR  CORN, GRAIN - ACRES HARVESTED  71,353,000
31613  1953   YEAR  CORN, GRAIN - ACRES HARVESTED  70,738,000
40954  1954   YEAR  CORN, GRAIN - ACRES HARVESTED  68,668,000

The column was dropped successfully.

That is the last of the filtering I need to do. However, the indeces are now a bit wonky. I wont to reset the indeces for ease of use.

In [17]:
# reset all indeces in corn_df_working
corn_df_working.reset_index(drop=True, inplace=True)

# confirm indeces reset successfully
display(corn_df_working.head())

Year Period                      Data Item       Value
0  1950   YEAR  CORN, GRAIN - ACRES HARVESTED  72,398,000
1  1951   YEAR  CORN, GRAIN - ACRES HARVESTED  71,191,000
2  1952   YEAR  CORN, GRAIN - ACRES HARVESTED  71,353,000
3  1953   YEAR  CORN, GRAIN - ACRES HARVESTED  70,738,000
4  1954   YEAR  CORN, GRAIN - ACRES HARVESTED  68,668,000

The indeces reset correctly.

## Clean Data Item Entries

All of the data items are for grain corn - the kind of corn you eat at the dinner table. Because of this, each data item contains the prefix phrase `"CORN, GRAIN - "`. To increase readability/understandability, I will remove this phrase from each data item. To do this, I will split the `Data Item` column in to two columns, `Corn, Grain` and `Data Item`, using "` - `" as the deliminator.

In [18]:
# split the "Data Item" column into two columns, seperating them at the proper string
corn_df_working[['Corn, Grain','Data Item']] = corn_df_working['Data Item'].str.split(' - ',expand=True)

# confirm that the data was split successfully
corn_df_working.sample(5)

Year Period                           Data Item       Value  Corn, Grain
1354  2011    MAR  PRICE RECEIVED, MEASURED IN $ / BU        5.53  CORN, GRAIN
786   1968    OCT  PRICE RECEIVED, MEASURED IN $ / BU       0.962  CORN, GRAIN
121   2012   YEAR                     ACRES HARVESTED   8,073,334  CORN, GRAIN
60    1997   YEAR                     ACRES HARVESTED  72,671,000  CORN, GRAIN
1447  2017    DEC  PRICE RECEIVED, MEASURED IN $ / BU        3.23  CORN, GRAIN

The column was properly split. I can now drop the `Corn, Grain` column, as it is contains no useful information.

In [19]:
# drop the "Corn, Grain" column
corn_df_working.drop(columns='Corn, Grain', inplace=True)

# confirm the drop was successful
corn_df_working.head(5)

Year Period        Data Item       Value
0  1950   YEAR  ACRES HARVESTED  72,398,000
1  1951   YEAR  ACRES HARVESTED  71,191,000
2  1952   YEAR  ACRES HARVESTED  71,353,000
3  1953   YEAR  ACRES HARVESTED  70,738,000
4  1954   YEAR  ACRES HARVESTED  68,668,000

The column was dropped successfully.

## Export compiled dataframe as a csv

The corn data frame is now ready for export. I am exporting this csv from this notebook and will seperate it into individual dataframes in another notebook so as to simplify the analysis process and seperate my steps.

In [20]:
corn_df_working.to_csv('./compiled_corn_df.csv')

This is the end of this notebook.